### PR TITLE
Following a text file

### PR DESCRIPTION
**Pull request by jglaser**
## Description
A new module to follow text files as they are being appended to.

## Motivation and Context
Allows the user to do the equivalent of `tail -f` in the web browser, to examine the output of running simulations for example

## Types of Changes

- [ ] Documentation update
- [ ] Bug fix
- [x] New feature
- [ ] Breaking change<sup>1</sup>

<sup>1</sup>The change breaks (or has the potential to break) existing functionality.

## Checklist:

- [x] I am familiar with the [**Contributing Guidelines**](https://github.com/glotzerlab/signac-dashboard/blob/master/CONTRIBUTING.md).
- [x] I agree with the terms of the [**Contributor Agreement**](https://github.com/glotzerlab/signac-dashboard/blob/master/ContributorAgreement.md).
- [x] My name is on the [list of contributors](https://github.com/glotzerlab/signac-dashboard/blob/master/contributors.txt).
- [x] My code follows the [code style guideline](https://github.com/glotzerlab/signac-dashboard/blob/master/CONTRIBUTING.md#code-style) of this project.
- [x] The changes introduced by this pull request are covered by existing or newly introduced tests.

If necessary:
- [ ] I have updated the API documentation as part of the package doc-strings.
- [ ] I have created a separate pull request to update the [framework documentation](https://docs.signac.io/) on [signac-docs](https://github.com/glotzerlab/signac-docs) and linked it here.
- [ ] I have updated the [changelog](https://github.com/glotzerlab/signac-dashboard/blob/master/changelog.txt).

### CODE REVIEW

**Comment by jglaser** [comment]
Should be merged after #82 

Screen Shot 2020-04-17 at 5 53 04 PM

**Comment by bdice** [comment]
@jglaser Awesome. Thanks for the pull requests. I might review this PR and #82 together, for simplicity. Both features look close to complete. Is it feasible to spawn a separate process for matplotlib, to avoid the lack of thread safety? Making Flask operate all requests on a single thread is very bad when you have many jobs on the same page — webservers are almost always meant to be run multithreaded.

**Comment by jglaser** [comment]
I believe I found a solution that involves `multiprocessing`, the (thread-safe) `Queue` object and a lock. It does appear to be slower than the straight forward single threaded implementation though, due to the overhead of communicating between tasks. Still, the performance issue is localized to the `plotter` module, and therefore other modules run in multi-threaded manner as before. Going forward it may be worthwhile evaluating alternatives to `flask` as a web server. I believe it interoperates with other servers and its own `http` server component is only intended as a development purposes anyway.

**Comment by bdice** [comment]
@jglaser This seems like a good solution with multiprocessing, that was what I had in mind. Try the WSGI examples here if you have performance issues: https://github.com/glotzerlab/signac-dashboard/tree/master/examples/wsgi

Flask is quite performant if you use a “real” server like gunicorn. There are many WSGI servers, but I have found gunicorn to be a great balance of performance and ease of setup for signac-dashboard applications.

**Comment by bdice** [comment]
@jglaser A couple questions:
1. Would it be better to provide a callable `create_figure(job)` to the Plotter constructor instead of making a subclass? That’s how the TextDisplay works. https://docs.signac.io/projects/dashboard/en/latest/api.html

2. Should we avoid the pyplot interface and use the object oriented API?

Also please make the style checks pass.

**Comment by jglaser** [comment]
Yes, I am now using a `callable` argument. Regarding the object oriented API, what specifically did you have in mind? I did not find a way to close a figure without `plt.close()`, and if you don't explicitly close a matplotlib figure they will accumulate and mpl will issue a warning about too many open figures.

**Comment by jglaser** [comment]
I have a few possible design improvements in mind:

- use a nice switch instead of a checkbox for the `follow` button. Like here: https://wikiki.github.io/form/switch/ . What would be the best way to include those CSS templates?
- In the Statepoint Viewer, use a table layout for nicer organization

But these are optional and we can leave that to later PRs.

**Comment by jglaser** [comment]
there is an issue with testing, where the instantiation of a module requires arguments for which no default is set. How would you proceed here?

**Comment by bdice** [comment]
@jglaser Maybe use None as a default argument and show an error message if the filename is None? I’m replying from my phone right now, sorry if this is missing something obvious.

**Comment by jglaser** [comment]
Note that I added the context manager to the module. If you think that makes it inconsistent between modules, as `Plotter` requires one, whereas the others do not, we could make `dashboard` itself be a context manager and forward entry and exit points to the modules...

**Comment by bdice** [comment]
@jglaser The current version of the code doesn't work for me. I can't get the plot to render, it just loads forever. I see the `__exit__` method of the context manager is being called as soon as the context is closed, which occurs immediately after appending the module to the list. This doesn't seem like the intended behavior.

**Comment by jglaser** [comment]
fixed indentation. Try again?

File: signac_dashboard/templates/cards/logger.html [hunk 1]
```diff
@@ -0,0 +1,50 @@
+<div>
+    <input type="checkbox" name="follow" id="follow-{{jobid}}" onchange="logger_{{jobid}}.setFollow()">
```
Comment (bdice) [inline signac_dashboard/templates/cards/logger.html:2]: This won't work if multiple loggers are present. The ids should include the job id and some other unique identifier. Also see the examples of the Notes and DocumentEditor modules for how to include custom JavaScript. The JS shouldn't be included repeatedly for each job id / module instance.

**Comment by b-butler** [comment]
@jglaser, do you still plan to work on this PR?

**Comment by csadorf** [comment]
Closing for now, please feel free to reopen if there is renewed interest.

**Pull request closed without merging**